packages/@lwc/engine-dom/src/patches/global-registry.ts

```diff
@@ -0,0 +1,398 @@
+// attributeChangedCallback
+function patchAttributes(
+    instance: HTMLElement,
+    registeredDefinition: Definition,
+    instancedDefinition: Definition
+) {
+    const { observedAttributes, attributeChangedCallback } = instancedDefinition;
+    if (observedAttributes.size === 0 || isUndefined(attributeChangedCallback)) {
+        return;
+    }
+    const offset = getObservedAttributesOffset(registeredDefinition, instancedDefinition);
+    if (offset.size === 0) {
+        return;
+    }
+    // instance level patches
+    defineProperties(instance, {
+        setAttribute: {
+            value: function setAttribute(name: string, value: any) {
+                if (offset.has(name)) {
+                    const old = nativeGetAttribute.call(this, name);
+                    // maybe we want to call the super.setAttribute rather than the native one
+                    nativeSetAttribute.call(this, name, value);
+                    attributeChangedCallback!.call(this, name, old, value + '');
+                } else {
+                    nativeSetAttribute.call(this, name, value);
+                }
+            },
+            writable: true,
+            enumerable: true,
+            configurable: true,
+        },
+        removeAttribute: {
+            value: function removeAttribute(name: string) {
+                if (offset.has(name)) {
+                    const old = nativeGetAttribute.call(this, name);
+                    // maybe we want to call the super.removeAttribute rather than the native one
+                    nativeRemoveAttribute.call(this, name);
+                    attributeChangedCallback!.call(this, name, old, null);
+                } else {
+                    nativeRemoveAttribute.call(this, name);
+                }
+            },
+            writable: true,
+            enumerable: true,
+            configurable: true,
+        },
+    });
+}
+
+// User extends this HTMLElement, which returns the CE being upgraded
+let upgradingInstance: HTMLElement | undefined;
+
+// Helper to upgrade an instance with a CE definition using "constructor call trick"
+function internalUpgrade(
+    instance: HTMLElement,
+    registeredDefinition: Definition,
+    instancedDefinition: Definition
+) {
+    setPrototypeOf(instance, instancedDefinition.UserCtor.prototype);
+    definitionForElement.set(instance, instancedDefinition);
+    // attributes patches when needed
+    if (instancedDefinition !== registeredDefinition) {
+        patchAttributes(instance, registeredDefinition, instancedDefinition);
+    }
+    // Tricking the construction path to believe that a new instance is being created,
+    // that way it will execute the super initialization mechanism but the HTMLElement
+    // constructor will reuse the instance by returning the upgradingInstance.
+    // This is by far the most important piece of the puzzle
+    upgradingInstance = instance;
+    try {
```

Comment:
this try/catch seems unnecessary.

Reply:
According to the Karma coverage tests, the `catch` condition is actually hit in our tests. Although the `if` inside of the `catch` is never hit.

![Screen Shot 2022-06-03 at 9 28 44 AM](https://user-images.githubusercontent.com/283842/171906873-b1734a7f-416c-4a7f-b9e5-fd0379eebc5a.png)

Setting a breakpoint, it looks like this `catch` handles the scenario where the constructor throws an error. E.g. this test:

https://github.com/salesforce/lwc/blob/d93e2963d8d32b138ce9ed130d98e7bef84b3d74/packages/integration-karma/test/component/LightningElement.setAttribute/index.spec.js#L15-L22